Added email-domain flag to restrict access to given domain

keattang opened the pull request:
Specifying the --email-domain flag allows you to prevent users with emails that don't match the given domain from accessing your secrets. This allows you to more confidently and securely expose kuberos to the internet

This is my first time writing go so please excuse any obvious mistakes. Happy to accept/implement feedback

keattang committed:
Added email-domain flag to restrict access

keattang committed:
Added email-domain flag to restrict access

keattang committed:
Added email-domain flag to restrict access

keattang committed:
Added email-domain flag to restrict access

negz commented:
@keattang It's been a while, but I feel like I've managed to get this behaviour without code changes before - I think via a setting at Google's end of a Google OIDC setup. I imagine not all OIDC providers let you filter via domain, so this seems reasonable.

negz commented on README.md:
```diff
@@ -72,6 +72,8 @@ Flags:
                          Wait this long for sessions to end before shutting
                          down.
+      --email-domain     Restrict access to users with emails that match
+                         this domain
```
A nitpick, but I'd prefer this help output be created by running `kuberos --help` and pasting it in. It's already a bit stale so you'd be doing me a favour. :)

keattang replied:
Done :)

negz commented on extractor/oidc.go:
```diff
@@ -59,13 +61,13 @@ func Logger(l *zap.Logger) Option {
 
 // NewOIDC creates a new OIDC extractor.
-func NewOIDC(v *oidc.IDTokenVerifier, oo ...Option) (OIDC, error) {
+func NewOIDC(v *oidc.IDTokenVerifier, emailDomain string, oo ...Option) (OIDC, error) {
```
I've tried to stick to a pattern in which only mandatory arguments (for example the `*oidc.IDTokenVerifier`) are 'first class' arguments to `NewOIDC`, with optional arguments implemented as `Option`s. Would you be willing to take a shot at copying the `Option` pattern for the email domain option?

keattang replied:
It appears below that you are looping over the options and calling them as functions on the `oe` object. I wasn't quite sure how to unpick that. What would you suggest?

negz replied:
Yeah, that's the pattern I was suggesting. In this specific case you've already added `emailDomain` so the idea is to add an 'option' function that takes an `oidcExtractor` as an argument and sets `emailDomain`. There's a good article about the pattern [here](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

Something like:
```go
func EmailDomain(domain *string) Option {
	return func(o *oidcExtractor) error {
		o.emailDomain = domain
		return nil
	}
}
```

That way the following call will result in the default (i.e. an extractor `emailDomain` set to the empty string):
```go
e, err := extractor.NewOIDC(provider.Verifier(&oidc.Config{ClientID: *clientID}))
```

While this call will result in a custom email domain:
```go
e, err := extractor.NewOIDC(provider.Verifier(&oidc.Config{ClientID: *clientID}), extractor.EmailDomain("example.org"))
```

keattang committed:
Updated readme with --help text

negz commented on cmd/kuberos/kuberos.go:
```diff
@@ -48,6 +48,7 @@ func main() {
 		debug       = app.Flag("debug", "Run with debug logging.").Short('d').Bool()
 		extraScopes = app.Flag("extra-scopes", "List of additional scopes to provide in token.").Strings()
 		grace       = app.Flag("shutdown-grace-period", "Wait this long for sessions to end before shutting down.").Default("1m").Duration()
+		emailDomain = app.Flag("email-domain", "The eamil domain to restrict access to.").String()
```
Just noticed there's a typo in `email` here.

negz commented on extractor/oidc.go:
```diff
@@ -34,9 +35,10 @@ type OIDC interface {
+	log 		*zap.Logger
+	v   		*oidc.IDTokenVerifier
+	h   		*http.Client
+	emailDomain string
```
This looks like it needs `gofmt` run on it to align `string` with the above types.

keattang committed:
Code review changes

keattang committed:
Code review changes

keattang committed:
Code review changes

keattang committed:
Code review changes

keattang commented:
@negz I implemented the changes you suggested. I also rebased my branch on your master but github seems to be showing those changes as part of the PR for some reason. I can close this and open another PR so that you just see my changes if you'd like?

negz commented:
@keattang Apologies for the super slow reply!

I've often fixed issues similar to this by running (from your branch):
```
# You might want to replace origin with upstream to rebase on github.com/negz rather than your fork's master
git fetch origin
git rebase origin/master
git push --force
```

keattang commented:
@negz that seems to have done the trick

negz commented:
Merged! Thanks again, and sorry it took so long to merge this.

keattang commented:
Awesome thanks! Any idea on when that may be released to docker hub? Would love to get back on to the official release

negz commented:
@keattang It's there already; Travis pushes master to Docker Hub whenever a PR is merged. `negz/kuberos:ab5b206` should include your changes.

keattang commented:
great!
